src/jit/emitxarch.cpp

```diff
@@ -13663,6 +13663,1128 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
+            memAccessKind     = PERFSCORE_MEMORY_WRITE;
+            break;
+
+        case IF_MRW:
+            result.insLatency = PERFSCORE_LATENCY_RD_WR_CONST_ADDR;
+            memAccessKind     = PERFSCORE_MEMORY_READ_WRITE;
+            break;
+
+        // Model a read from memory location, possible def to use latency from L0 or L1 cache
+        case IF_ARD:
+            result.insLatency = PERFSCORE_LATENCY_RD_GENERAL;
+            memAccessKind     = PERFSCORE_MEMORY_READ;
+            break;
+
+        case IF_AWR:
+            result.insLatency = PERFSCORE_LATENCY_WR_GENERAL;
+            memAccessKind     = PERFSCORE_MEMORY_WRITE;
+            break;
+
+        case IF_ARW:
+            result.insLatency = PERFSCORE_LATENCY_RD_WR_GENERAL;
+            memAccessKind     = PERFSCORE_MEMORY_READ_WRITE;
+            break;
+
+        case IF_NONE:
+            result.insLatency = PERFSCORE_LATENCY_ZERO;
+            memAccessKind     = PERFSCORE_MEMORY_NONE;
+            break;
+
+        default:
+            assert(!"Unhandled insFmt for switch (memFmt)");
+            result.insLatency = PERFSCORE_LATENCY_ZERO;
+            memAccessKind     = PERFSCORE_MEMORY_NONE;
+            break;
+    }
+    result.insMemoryAccessKind = memAccessKind;
+
+    switch (ins)
+    {
+        case INS_nop:
+        case INS_int3:
+            assert(memFmt == IF_NONE);
+            result.insThroughput = PERFSCORE_THROUGHPUT_4X;
+            result.insLatency    = PERFSCORE_LATENCY_ZERO;
+            break;
+
+        case INS_push:
+        case INS_push_hide:
+            result.insThroughput = PERFSCORE_THROUGHPUT_1C;
+            if (insFmt == IF_RRD) // push  reg
+            {
+                // For pushes (stack writes) we assume that the full latency will be covered
+                result.insLatency = PERFSCORE_LATENCY_ZERO;
+            }
+            break;
+
+        case INS_pop:
+        case INS_pop_hide:
+            if (insFmt == IF_RWR) // pop   reg
+            {
+                result.insThroughput = PERFSCORE_THROUGHPUT_2X;
+                // For pops (stack reads) we assume that the full latency will be covered
+                result.insLatency = PERFSCORE_LATENCY_ZERO;
+            }
+            else
+            {
+                result.insThroughput = PERFSCORE_THROUGHPUT_1C;
+            }
+            break;
+
+        case INS_inc:
+        case INS_dec:
+        case INS_neg:
+        case INS_not:
+            if (memFmt == IF_NONE)
+            {
+                // ins   reg
+                result.insThroughput = PERFSCORE_THROUGHPUT_4X;
+                result.insLatency    = PERFSCORE_LATENCY_1C;
+            }
+            else
+            {
+                // ins   mem
+                result.insThroughput = PERFSCORE_THROUGHPUT_1C;
+            }
+            break;
+
+#ifdef _TARGET_AMD64_
+        case INS_movsxd:
+#endif
+        case INS_mov:
+        case INS_movsx:
+        case INS_movzx:
```

Comment:
Should it be modeled that `movzx` can be frequently elided in the register renamer? (https://www.agner.org/optimize/microarchitecture.pdf)

Reply:
For a `mov/movxz reg, reg`  we will return:

    result.insThroughput = PERFSCORE_THROUGHPUT_4X;
    result.insLatency = PERFSCORE_LATENCY_ZERO;

Which results in a net cost of 0.25 for this instruction. 
(which is the minimal cost of any instruction that we model)
If the instruction reads from memory we will always have to pay a memory read latency cost.

Reply:
> result.insLatency = PERFSCORE_LATENCY_ZERO;

This seems incorrect. You said that this was modeled after Skylake but this CPU does not eliminate `movzx`.

Reply:
I can change this to a 1 cycle latency, but due to the current heuristics that won't actually change the PerfScore that we calculate.  (I.e. one cycle of latency for an instruction is currently equivalent to a zero latency instruction,  Note that the issue throughput can never be zero so it's cost will be used)

The current heuristic assumes that the processor will be always be able to eliminate one cycle of latency.  We would need to accurately model the use/defs of registers and also simulate execution to do better.